File: app/src/main/java/com/pawels96/skyrimperkcalculator/presentation/common/views/LoopingViewPager.java
package com.pawels96.skyrimperkcalculator.presentation.common.views;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;

import android.util.AttributeSet;

/**
 * ViewPager which allows for moving between the first and last item
 */

public class LoopingViewPager extends ViewPager {
    public LoopingViewPager(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public interface MovementListener {
        void onActionDown();
    }

    private MovementListener listener;

    public void setListener(MovementListener listener) {
        this.listener = listener;
    }

    public void enableLoop(final int pages) {

        addOnPageChangeListener(new ViewPager.OnPageChangeListener() {

            int previousState, currentState;

            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) { }

            @Override
            public void onPageSelected(int position) { }

            @Override
            public void onPageScrollStateChanged(int state) {

                if (state == 1)
                    listener.onActionDown();

                int currentPage = getCurrentItem();
                if (currentPage == pages - 1 || currentPage == 0) {
                    previousState = currentState;
                    currentState = state;
                    if (previousState == 1 && currentState == 0) {
                        setCurrentItem(currentPage == 0 ? pages - 1 : 0, false);
                    }
                }
            }
        });
    }
}

File: app/src/main/java/com/pawels96/skyrimperkcalculator/presentation/common/views/LoopingViewPager.kt
package com.pawels96.skyrimperkcalculator.presentation.common.views

import android.content.Context
import android.util.AttributeSet
import androidx.viewpager.widget.ViewPager

/**
 * ViewPager which allows for moving between the first and last item
 */
class LoopingViewPager(context: Context, attrs: AttributeSet?) : ViewPager(context, attrs) {

    var onScrollStarted: (() -> Unit)? = null

    fun enableLoop(pages: Int) {
        addOnPageChangeListener(object : OnPageChangeListener {
            var previousState = 0
            var currentState = 0

            override fun onPageScrolled(
                position: Int,
                positionOffset: Float,
                positionOffsetPixels: Int
            ) = Unit

            override fun onPageSelected(position: Int) = Unit

            override fun onPageScrollStateChanged(state: Int) {
                if (state == 1) {
                    onScrollStarted?.invoke()
                }

                val currentPage = currentItem
                if (currentPage == pages - 1 || currentPage == 0) {
                    previousState = currentState
                    currentState = state
                    if (previousState == 1 && currentState == 0) {
                        setCurrentItem(if (currentPage == 0) pages - 1 else 0, false)
                    }
                }
            }
        })
    }
}